docs/architecture/deep-dives/state/index.md

```diff
@@ -0,0 +1,379 @@
+`KeyDefinition` classes that allow for reasonable reuse of partial namespaces while allowing
+expansion to precise keys. They exist to help minimize the potential of overlaps in a distributed
+storage framework.
+
+:::warning
+
+Once you have created the definitions, you need to take extreme caution when changing any part of
+the namespace. If you change the name of a `StateDefinition` pointing at `"disk"` without also
+migrating data from the old name to the new name you will lose data. Data pointing at `"memory"` can
+have their name changed.
+
+:::
+
+#### `StateDefinition`
+
+`StateDefinition` is a simple API but a very core part of making the State Provider Framework work
+smoothly. It defines a storage location and top-level namespace for storage. Teams will interact
+with it only in a single, `state-definitions.ts`, file in the
+[`clients`](https://github.com/bitwarden/clients) repository. This file is located under platform
+code ownership but teams are expected to create edits to it. A team will edit this file to include a
+line like such:
+
+```typescript
+export const MY_DOMAIN_DISK = new StateDefinition("myDomain", "disk");
+```
+
+The first argument to the `StateDefinition` constructor is expected to be a human readable,
+camelCase formatted name for your domain, or state area. The second argument will either be the
+string literal `"disk"` or `"memory"` dictating where all the state using this `StateDefinition`
+should be stored. The platform team will be responsible to reviewing all new and updated entries in
+this file and will be looking to make sure that there are no duplicate entries containing the same
+state name and state location. Teams CAN have the same state name used for both `"disk"` and
+`"memory"` locations. Tests are included to ensure this uniqueness and core naming guidelines so you
+can ensure a review for a new `StateDefinition` entry can be done promptly and with very few
+surprises.
+
+_TODO: Make tests_
+
+:::note
+
+Secure storage is not currently supported as a storage location in the State Provider Framework. For
+now, don't migrate data that is stored in secure storage but please contact platform when you have
+data you wanted to migrate so we can prioritize a long term solution. If you need new data in secure
+storage, use `StateService` for now.
+
+:::
+
+#### `KeyDefinition`
+
+`KeyDefinition` builds on the idea of [`StateDefinition`](#statedefinition) but it gets more
+specific about the data to be stored. `KeyDefinition`s can also be instantiated in your own teams
+code. This might mean creating it in the same file as the service you plan to consume it or you may
+want to have a single `key-definitions.ts` file that contains all the entries for your team. Some
+example instantiations are:
+
+```typescript
+const MY_DOMAIN_DATA = new KeyDefinition<MyState>(MY_DOMAIN_DISK, "data", {
+  deserializer: (jsonData) => null, // convert to your data from json
```

Comment:
💭 I found this a bit misleading. `jsonData` is already parsed, whereas I expect JSON to be the stringified version of data. Had there been an implementation provided or there were a type annotation or something, I'd not have been as confused.

Example:
```typescript
const MY_DOMAIN_DATA = new KeyDefinition<MyState>(MY_DOMAIN_DISK, "data", {
  // convert to your data from serialized representation `{ foo: string }` 
  // to a fully-typed `MyState`.
  deserializer: (jsonData) => MyState.fromJson(jsonData), 
});
```

Reply:
Very, very fair. I took this exact snippet and committed it, thanks!